File: test/data/cpp/test05/BaseState.cpp
#include "BaseState.h"
BaseState::BaseState(IControllee* controllee)
{
    _controllee = controllee;
}
BaseState::~BaseState()
{
}
void BaseState::Setup(bool resume, bool deepResume)
{
}
BaseState* BaseState::TransitEvent2()
{
    _controllee->NoTransition(GetStateName(), "Event2");
    return this;
}
BaseState* BaseState::TransitEvent21()
{
    _controllee->NoTransition(GetStateName(), "Event21");
    return this;
}
BaseState* BaseState::TransitEvent31()
{
    _controllee->NoTransition(GetStateName(), "Event31");
    return this;
}
BaseState* BaseState::TransitEvent42()
{
    _controllee->NoTransition(GetStateName(), "Event42");
    return this;
}
BaseState* BaseState::TransitEvent421()
{
    _controllee->NoTransition(GetStateName(), "Event421");
    return this;
}
BaseState* BaseState::TransitEvent431()
{
    _controllee->NoTransition(GetStateName(), "Event431");
    return this;
}
BaseState* BaseState::TryTransitWithoutEvent()
{
    return this;
}
void BaseState::SetupSubState(BaseState* child, bool resume)
{
    if(!resume)
    {
        _currentSubState = child;
    }
    if(_currentSubState != 0)
    {
        _currentSubState->Setup(false, false);
    }
}
BaseState* BaseState::CurrentSubState()
{
    return _currentSubState;
}
BaseState* BaseState::TransitBySubState(BaseState* nextState)
{
    if(nextState == 0 || _currentSubState == 0)
    {
        return nextState;
    }
    BaseState* parentOfNextState = _currentSubState->GetParent();
    BaseState* parentOfCurrentState = nextState->GetParent();
    if(parentOfNextState != 0 && parentOfCurrentState != 0 && parentOfNextState == parentOfCurrentState)
    {
        _currentSubState = nextState;
        return this;
    }
    return nextState;
}
BaseState* BaseState::TransitForChild(BaseState* child)
{
    _currentSubState = child;
    BaseState* parent = GetParent();
    if(parent != 0)
    {
        return parent->TransitForChild(this);
    }
    return this;
}
BaseState* BaseState::OutlineState()
{
    BaseState* parent = GetParent();
    if(parent != 0)
    {
        return parent->TransitForChild(this);
    }
    return this;
}

File: test/data/typescript/test05/BaseState.ts
import {IControllee} from "./IControllee"
export abstract class BaseState
{
    private _controllee: IControllee;
    private _currentSubState: BaseState | null;
    public constructor(controllee: IControllee)
    {
        this._controllee = controllee;
        this._currentSubState = null;
    }
    public Setup(resume: boolean, deepResume: boolean): void
    {
        return;
    }
    public TransitEvent2(): BaseState | null
    {
        this._controllee.NoTransition(this.GetStateName(), "Event2");
        return this;
    }
    public TransitEvent21(): BaseState | null
    {
        this._controllee.NoTransition(this.GetStateName(), "Event21");
        return this;
    }
    public TransitEvent31(): BaseState | null
    {
        this._controllee.NoTransition(this.GetStateName(), "Event31");
        return this;
    }
    public TransitEvent42(): BaseState | null
    {
        this._controllee.NoTransition(this.GetStateName(), "Event42");
        return this;
    }
    public TransitEvent421(): BaseState | null
    {
        this._controllee.NoTransition(this.GetStateName(), "Event421");
        return this;
    }
    public TransitEvent431(): BaseState | null
    {
        this._controllee.NoTransition(this.GetStateName(), "Event431");
        return this;
    }
    public TryTransitWithoutEvent(): BaseState
    {
        return this;
    }
    public SetupSubState(child: BaseState, resume: boolean): void
    {
        if(!resume)
        {
            this._currentSubState = child;
        }
        if(this._currentSubState != null)
        {
            this._currentSubState.Setup(false, false);
        }
    }
    public CurrentSubState(): BaseState | null
    {
        return this._currentSubState;
    }
    public TransitBySubState(nextState: BaseState | null): BaseState | null
    {
        if(nextState == null || this._currentSubState == null)
        {
            return nextState;
        }
        let parentOfNextState: BaseState | null = this._currentSubState.GetParent();
        let parentOfCurrentState: BaseState | null = nextState.GetParent();
        if(parentOfNextState != null && parentOfCurrentState != null && parentOfNextState == parentOfCurrentState)
        {
            this._currentSubState = nextState;
            return this;
        }
        return nextState;
    }
    public TransitForChild(child: BaseState): BaseState
    {
        this._currentSubState = child;
        let parent: BaseState | null = this.GetParent();
        if(parent != null)
        {
            return parent.TransitForChild(this);
        }
        return this;
    }
    public OutlineState(): BaseState
    {
        let parent: BaseState | null = this.GetParent();
        if(parent != null)
        {
            return parent.TransitForChild(this);
        }
        return this;
    }
    public ResumeSubState(subState: BaseState | null): void
    {
        this._currentSubState = subState;
    }
    public abstract GetStateName(): string;
    public abstract GetParent(): BaseState | null;
    public abstract GetStateID(): number;
}